PathTracking/pure_pursuit/pure_pursuit.py

```diff
@@ -31,89 +30,106 @@ def __init__(self, x=0.0, y=0.0, yaw=0.0, v=0.0):
         self.rear_x = self.x - ((L / 2) * math.cos(self.yaw))
         self.rear_y = self.y - ((L / 2) * math.sin(self.yaw))
 
+    def update(self, a, delta):
 
-def update(state, a, delta):
+        self.x += self.v * math.cos(self.yaw) * dt
+        self.y += self.v * math.sin(self.yaw) * dt
+        self.yaw += self.v / L * math.tan(delta) * dt
+        self.v += a * dt
+        self.rear_x = self.x - ((L / 2) * math.cos(self.yaw))
+        self.rear_y = self.y - ((L / 2) * math.sin(self.yaw))
 
-    state.x = state.x + state.v * math.cos(state.yaw) * dt
-    state.y = state.y + state.v * math.sin(state.yaw) * dt
-    state.yaw = state.yaw + state.v / L * math.tan(delta) * dt
-    state.v = state.v + a * dt
-    state.rear_x = state.x - ((L / 2) * math.cos(state.yaw))
-    state.rear_y = state.y - ((L / 2) * math.sin(state.yaw))
+    def calc_distance(self, point_x, point_y):
 
-    return state
+        dx = self.rear_x - point_x
+        dy = self.rear_y - point_y
+        return math.hypot(dx, dy)
 
 
-def PIDControl(target, current):
-    a = Kp * (target - current)
+class States:
 
-    return a
+    def __init__(self):
+        self.x = []
+        self.y = []
+        self.yaw = []
+        self.v = []
+        self.t = []
 
+    def append(self, t , state):
+        self.x.append(state.x)
+        self.y.append(state.y)
+        self.yaw.append(state.yaw)
+        self.v.append(state.v)
+        self.t.append(t)
 
-def pure_pursuit_control(state, cx, cy, pind):
 
-    ind = calc_target_index(state, cx, cy)
+def PIDControl(target, current):
+    a = Kp * (target - current)
 
-    if pind >= ind:
-        ind = pind
+    return a
 
-    if ind < len(cx):
-        tx = cx[ind]
-        ty = cy[ind]
-    else:
-        tx = cx[-1]
-        ty = cy[-1]
-        ind = len(cx) - 1
 
-    alpha = math.atan2(ty - state.rear_y, tx - state.rear_x) - state.yaw
+class Trajectory:
+    def __init__(self, cx, cy):
+        self.cx = cx
+        self.cy = cy
+        self.old_nearest_point_index = None
 
-    Lf = k * state.v + Lfc
+    def __call__(self, state):
```

Comment:
I think this function should has a function name instead of __call__ like "search_target_index" for readability.

Reply:
I have renamed it `search_target_index` as it was originally.